### ISSUE TITLE
S3_SKIP_SIGNATURE_VALIDATION does not work

### ISSUE THREAD

**Issue by sergeyfsv-da** [localstack/localstack#4160]
# Type of request: S3_SKIP_SIGNATURE_VALIDATION does not and files uploading still fails

[ *] bug report

# Detailed description

Despite the fact that env variable is provided for skipping S3 presigned url validation, uploading of files still fails with the signature in v2 format.

How is that possible to get "Presign signature calculation failed" if the logic in the code is the following. So, with the S3_SKIP_SIGNATURE_VALIDATION env variable set, response should *always* be None

    elif is_v4 and not is_v2:
        response = authenticate_presign_url_signv4(method, path, headers, data, url, query_params, request_dict)

    if response is not None:
        LOGGER.error('Presign signature calculation failed: %s' % response)
        return response

## Proposed Pull-Request

https://github.com/localstack/localstack/pull/4161

## Expected behavior

Get http status code 200

## Actual behavior

Get 403/

localstack_1     | 2021-06-17T17:32:10:DEBUG:localstack.services.s3.s3_utils: Received presign S3 URL: http://localhost:4566/evidence.api.lambda.s3.bucket/file.doc?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=test%2F20210617%2Feu-west-1%2Fs3%2Faws4_request&X-Amz-Date=20210617T173155Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host&X-Amz-Signature=b3e88a9dcfa4adb1ffbfb245d6eed38a1a3fa2b5e23f1f543a19201f1124ddae
localstack_1     | 2021-06-17T17:32:10:WARNING:localstack.services.s3.s3_utils: Signatures do not match, but not raising an error, as S3_SKIP_SIGNATURE_VALIDATION=1
localstack_1     | 2021-06-17T17:32:10:ERROR:localstack.services.s3.s3_utils: Presign signature calculation failed: <Response [403]>

# Steps to reproduce

docker-compose.yml

  localstack:
    image: localstack/localstack:0.12.12
    environment:
      - HOST_TMP_FOLDER=c:/xxx/backend/@docker/localstack
      - DATA_DIR=/tmp/localstack/data
      - LAMBDA_EXECUTOR=docker
      - LAMBDA_REMOTE_DOCKER=false
      - MAIN_CONTAINER_NAME=backend_localstack_1
      - LAMBDA_DOCKER_FLAGS=-e AWS_ACCESS_KEY_ID="test" -e AWS_SECRET_ACCESS_KEY="test" -e AWS_DEFAULT_REGION="eu-west-1"
      - DEBUG=1
      - S3_SKIP_SIGNATURE_VALIDATION=1
      - AWS_ACCESS_KEY_ID=test
      - AWS_SECRET_ACCESS_KEY=test
      - AWS_DEFAULT_REGION=eu-west-1
      - SERVICES=iam,lambda,apigateway,cloudformation,sns,sqs,sts,ssm,dynamodb,s3,ecr
      - DOCKER_HOST=unix:///var/run/docker.sock
    volumes:
       - "./@docker/localstack:/tmp/localstack"
       - /var/run/docker.sock:/var/run/docker.sock
    ports:
      - 443:443
      - 4566:4566

Prepare the localstack env:
aws s3 mb s3://evidence.api.lambda.s3.bucket --region eu-west-1 --endpoint http://localhost:4566
aws s3 presign s3://evidence.api.lambda.s3.bucket/file.doc -endpoint http://localhost:4566 --region eu-west-1

Create any file.doc

curl -i -X PUT http://localhost:4566/evidence.api.lambda.s3.bucket/file.doc?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=test%2F20210617%2Feu-west-1%2Fs3%2Faws4_request&X-Amz-Date=20210617T141615Z&X-Amz-Expires=604800&X-Amz-SignedHeaders=host&X-Amz-Signature=f3ef5f3d9c906da7be179a8d182bb24a61376b5f9eb3cc1198d200b9651b85a4 -H 'content-type: application/x-www-form-urlencoded' --upload-file file.doc

## Command used to start LocalStack

docker-compose up localstack

## That is not relevant to this bug, but the signature generated by the code snippet below does not pass signature validation:

            services.AddAWSService<IAmazonS3>();
            var s3Client = new AmazonS3Client(
                new BasicAWSCredentials("test", "test"),
                new AmazonS3Config
                {
                    ServiceURL = configuration.GetValue<string>("AWS:ServiceUrl"),
                    ForcePathStyle = true,
                    AuthenticationRegion = configuration.GetValue<string>("AWS:Region"),
                }
            );

            services.AddSingleton<IAmazonS3>(s3Client);

        private string GeneratePreSignedURL(string bucketName, string key)
        {
            var request = new GetPreSignedUrlRequest
            {
                BucketName = bucketName,
                Key        = key,
                Expires    = DateTime.UtcNow.AddHours(24.0),
                Verb       = HttpVerb.PUT,
            };

            return _s3Client.GetPreSignedURL(request);
        }

Dependencies

```
        <PackageReference Include="Amazon.Lambda.AspNetCoreServer" Version="6.0.2" />
        <PackageReference Include="AWSSDK.Extensions.NETCore.Setup" Version="3.7.0.1" />
        <PackageReference Include="AWSSDK.S3" Version="3.7.1.2" />
        <PackageReference Include="System.IO.Pipelines" Version="5.0.1" />

```

**Comment by macnev2013** [comment]
Hi @thrau , @whummer . Apologies for the delay. This is fixed in the latest version of `localstack`.

Tested with:
```
awslocal s3 mb s3://test
awslocal s3 cp test s3://test
awslocal s3 presign s3://test/test
```

Opening URL generated by presign into browser.

Can you give it a try ?

Thanks :)

**Comment by naspeh** [comment]
Hi @sergeyfsv-da it was reported a while ago, we have had a fix for expiration time #4164 since then.

Also, I was trying to reproduce this issue locally and found that [`aws s3 presign` generates url for GET method](https://docs.aws.amazon.com/cli/latest/reference/s3/presign.html#description) and you can't generate presign url for PUT method using `aws cli`, I have tested these steps (you can copy and run it as is)
```sh
bucket='evidence.api.lambda.s3.bucket'
file='file.doc'
awslocal s3 mb s3://$bucket

cat <<EOF | python | read upload_url
# generate presigned url for put_object (can't generate it by aws cli)
# instead of "awslocal s3 presign s3://evidence.api.lambda.s3.bucket/file.doc --region eu-west-1"

import os
import boto3

os.environ['AWS_DEFAULT_REGION'] = 'us-east-1'
os.environ['AWS_ACCESS_KEY_ID'] = 'test'
os.environ['AWS_SECRET_ACCESS_KEY'] = 'test'
url = boto3.client('s3', endpoint_url='http://localhost:4566').generate_presigned_url(
   ClientMethod='put_object',
   Params={'Bucket': '$bucket', 'Key': '$file'},
   ExpiresIn=3600
)
print(url)
EOF
echo "$upload_url"

touch "/tmp/$file"
curl -i -X PUT "$upload_url" -H 'content-type:application/x-www-form-urlencoded' --upload-file "/tmp/$file"
```

And it worked perfectly fine with output like this in server logs
```
2021-11-19T15:36:01.651:DEBUG:localstack.services.edge: IN(s3): "PUT /evidence.api.lambda.s3.bucket/file.doc?AWSAccessKeyId=test&Signature=76wN9LQBF0qhRj89lOumQADLvTE%3D&Expires=1637339761" - headers: {'Remote-Addr': '127.0.0.1', 'Host': 'localhost:4566', 'User-Agent': 'curl/7.79.1', 'Accept': '*/*', 'Content-Type': 'application/x-www-form-urlencoded', 'Content-Length': '0', 'X-Forwarded-For': '127.0.0.1, localhost:4566', 'x-localstack-edge': 'http://localhost:4566'} - data: b''
2021-11-19T15:36:01.652:DEBUG:localstack.services.s3.s3_utils: Received presign S3 URL: http://localhost:4566/evidence.api.lambda.s3.bucket/file.doc?AWSAccessKeyId=test&Signature=76wN9LQBF0qhRj89lOumQADLvTE%3D&Expires=1637339761
2021-11-19T15:36:01.652:DEBUG:localstack.services.s3.s3_utils: Valid presign url.
2021-11-19T15:36:01.656:DEBUG:localstack.services.edge: OUT(s3): "PUT /evidence.api.lambda.s3.bucket/file.doc?AWSAccessKeyId=test&Signature=76wN9LQBF0qhRj89lOumQADLvTE%3D&Expires=1637339761" - status: 200 - response headers: {'ETag': '"d41d8cd98f00b204e9800998ecf8427e"', 'last-modified': 'Fri, 19 Nov 2021 15:36:01 GMT', 'Content-Length': '0', 'x-amzn-requestid': 'MNNTLSZKQCTJXXLF8NU66BPS5KFLIW5YBK1YWIQ4S6WHCLFRSLKJ', 'Content-Type': 'text/html; charset=utf-8', 'Access-Control-Allow-Origin': '*', 'Server': 'Werkzeug/2.0.2 Python/3.9.2', 'Date': 'Fri, 19 Nov 2021 15:36:01 GMT', 'Location': 'http://evidence.api.lambda.s3.bucket.s3.localhost.localstack.cloud:4566/', 'x-amz-request-id': '35B8D9FC29476D98', 'x-amz-id-2': 'MzRISOwyjmnup35B8D9FC29476D987/JypPGXLh0OVFGcJaaO3KW/hRAqKOpIEEp'} - response:
```

I also changed `AWS_DEFAULT_REGION` to `us-east-2` to test V4 version
```
# change one line from snipped above
os.environ['AWS_DEFAULT_REGION'] = 'us-east-2'

# server logs
2021-11-19T15:40:44.067:DEBUG:localstack.services.edge: IN(s3): "PUT /evidence.api.lambda.s3.bucket/file.doc?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=test%2F20211119%2Fus-east-2%2Fs3%2Faws4_request&X-Amz-Date=20211119T154044Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host&X-Amz-Signature=bdbde6ad56d1b6c71737bcd8605753699ea8f744dbd17646f92e9dad2031f38c" - headers: {'Remote-Addr': '127.0.0.1', 'Host': 'localhost:4566', 'User-Agent': 'curl/7.79.1', 'Accept': '*/*', 'Content-Type': 'application/x-www-form-urlencoded', 'Content-Length': '0', 'X-Forwarded-For': '127.0.0.1, localhost:4566', 'authorization': 'AWS4-HMAC-SHA256 Credential=test/20211119/us-east-2/s3/aws4_request, SignedHeaders=host, Signature=bdbde6ad56d1b6c71737bcd8605753699ea8f744dbd17646f92e9dad2031f38c', 'x-localstack-edge': 'http://localhost:4566'} - data: b''
2021-11-19T15:40:44.067:DEBUG:localstack.services.s3.s3_utils: Received presign S3 URL: http://localhost:4566/evidence.api.lambda.s3.bucket/file.doc?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=test%2F20211119%2Fus-east-2%2Fs3%2Faws4_request&X-Amz-Date=20211119T154044Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host&X-Amz-Signature=bdbde6ad56d1b6c71737bcd8605753699ea8f744dbd17646f92e9dad2031f38c
2021-11-19T15:40:44.068:DEBUG:localstack.services.s3.s3_utils: Valid presign url.
2021-11-19T15:40:44.072:DEBUG:localstack.services.edge: OUT(s3): "PUT /evidence.api.lambda.s3.bucket/file.doc?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=test%2F20211119%2Fus-east-2%2Fs3%2Faws4_request&X-Amz-Date=20211119T154044Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host&X-Amz-Signature=bdbde6ad56d1b6c71737bcd8605753699ea8f744dbd17646f92e9dad2031f38c" - status: 200 - response headers: {'ETag': '"d41d8cd98f00b204e9800998ecf8427e"', 'last-modified': 'Fri, 19 Nov 2021 15:40:44 GMT', 'Content-Length': '0', 'x-amzn-requestid': 'YXA73HIH0PRXN3H2PL8UNJJZMYBO5UFXYPJ0R0SC0XPTOGOC9SJP', 'Content-Type': 'text/html; charset=utf-8', 'Access-Control-Allow-Origin': '*', 'Server': 'Werkzeug/2.0.2 Python/3.9.2', 'Date': 'Fri, 19 Nov 2021 15:40:44 GMT', 'Location': 'http://evidence.api.lambda.s3.bucket.s3.localhost.localstack.cloud:4566/', 'x-amz-request-id': '2A99B857914209C5', 'x-amz-id-2': 'MzRISOwyjmnup2A99B857914209C57/JypPGXLh0OVFGcJaaO3KW/hRAqKOpIEEp'} - response:
2021-11-19T15:40:44.072:DEBUG:localstack.services.edge: OUT(s3): "PUT /evidence.api.lambda.s3.bucket/file.doc?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=test%2F20211119%2Fus-east-2%2Fs3%2Faws4_request&X-Amz-Date=20211119T154044Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host&X-Amz-Signature=bdbde6ad56d1b6c71737bcd8605753699ea8f744dbd17646f92e9dad2031f38c" - status: 200 - response headers: {'ETag': '"d41d8cd98f00b204e9800998ecf8427e"', 'last-modified': 'Fri, 19 Nov 2021 15:40:44 GMT', 'Content-Length': '0', 'x-amzn-requestid': 'YXA73HIH0PRXN3H2PL8UNJJZMYBO5UFXYPJ0R0SC0XPTOGOC9SJP', 'Content-Type': 'text/html; charset=utf-8', 'Access-Control-Allow-Origin': '*', 'Server': 'Werkzeug/2.0.2 Python/3.9.2', 'Date': 'Fri, 19 Nov 2021 15:40:44 GMT', 'Location': 'http://evidence.api.lambda.s3.bucket.s3.localhost.localstack.cloud:4566/', 'x-amz-request-id': '2A99B857914209C5', 'x-amz-id-2': 'MzRISOwyjmnup2A99B857914209C57/JypPGXLh0OVFGcJaaO3KW/hRAqKOpIEEp'} - response:
```

Also, one note
> Generally, the assumption is that access key and secret key are both set to the value `test` for your client which generates the presigned URL. 

This is why I have this in my snipped above:
```py
os.environ['AWS_ACCESS_KEY_ID'] = 'test'
os.environ['AWS_SECRET_ACCESS_KEY'] = 'test'
```

I'm going to close this issue for now, but if you could test once again with the latest `localstack` image and still can reproduce the error please reopen the issue or create a new one.
